movementX and movementY are missing from pointerMove events (255207)

aprotyas opened the pull request:
#### fa2caa624d631c7085d6cb55a6b83ebc1c2ffe49
<pre>
movementX and movementY are missing from pointerMove events (255207)
rdar://108112600
<a href="https://bugs.webkit.org/show_bug.cgi?id=255207">https://bugs.webkit.org/show_bug.cgi?id=255207</a>

Reviewed by Wenson Hsieh.

This commit performs a few things:

- Expose `movementX`/`movementY` *generally* in MouseEvent structs
  rather than locking it behind POINTER_LOCK support. This is because
  even without implementing pointer lock on iOS, we can (a) easily
  compute these values and (b) the attributes are generally useful.

- Introduce a bunch of plumbing to pass touch event deltas from our
  EventHandler to the PointerEvent creation. The bookkeeping required to
  compute these touch deltas is placed in WebKitAdditions.

- Tangentially, changes the `PointerID` type to represent uint32_t
  instead of its signed sibling. I&apos;ve done this because we&apos;re passing in
  unsigned identifiers when creating PointerEvents, and I&apos;ve logged a
  few cases of ID *overflow*.

- Add a layout test to verify correctness of the movementX/Y fields of
  pointer events generated by touch gestures.

- Corrects some test expectations now that we expose
  `movementX`/`movementY` regardless of pointer lock support.

* LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields-expected.txt: Added.
* LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields.html: Added.
* LayoutTests/platform/gtk/pointer-lock/mouse-event-delivery-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/pointerevents/pointerlock/pointerevent_movementxy_with_pointerlock-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/pointerlock/movementX_Y_basic-expected.txt:
* Source/WebCore/dom/MouseEvent.idl:
* Source/WebCore/dom/MouseEventInit.idl:
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::create):
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::clear):
(WebCore::EventHandler::handleTouchEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/platform/PointerID.h:
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/263829@main">https://commits.webkit.org/263829@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50284229ed8dc084c4f974c36cbae90493e8aef3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7375 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6206 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5816 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7996 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5267 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7430 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5244 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5314 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5323 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5767 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4737 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5209 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5174 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1382 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5570 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -38,6 +38,7 @@
 #include "TextGranularity.h"
 #include "Timer.h"
 #include <memory>
+#include <utility>
```
is this import needed?

aprotyas replied:
I'm importing this for `std::pair`. I reckon `<utility>` can be removed safely because it gets transitively included from elsewhere, but I prefer to include-what-I-use ([why IYWU](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md)) in headers at least... is this against our preferred style? I can remove it then!

rr-codes commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -655,6 +655,7 @@ class EventHandler {
 
     TouchArray m_touches;
     RefPtr<LocalFrame> m_touchEventTargetSubframe;
+    HashMap<unsigned, std::pair<IntPoint, IntPoint>, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>> m_touchLastGlobalPositionAndDeltaMap;
```
are the default HashMap template arguments not sufficient?

aprotyas replied:
No, we need this because there can be touch IDs (map key) with value `0`, and that corrupts the hash table.

aprotyas replied:
Well, we really only need `WTF::UnsignedWithZeroKeyHashTraits<unsigned>`, but the `IntHash<>` template arg gets in the way 😛

whsieh commented on LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields.html:
```diff
@@ -0,0 +1,61 @@
+        function runTest()
+        {
+            window.addEventListener("pointerdown", event => {
+                debug("pointerup fired.");
```
This doesn't look right — do you mean `pointerdown fired.`?

aprotyas replied:
Yes, good catch. Fixed in https://github.com/WebKit/WebKit/commit/eed593fcf230c5752fc982cef0b7cbdda782897a

whsieh commented on LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields-expected.txt:
```diff
@@ -0,0 +1,31 @@
+
+
+
+    pointerup fired.
```
This should be `pointerdown`

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/eed593fcf230c5752fc982cef0b7cbdda782897a

whsieh commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -229,7 +229,7 @@ class EventHandler {
 
 #if ENABLE(IOS_TOUCH_EVENTS)
-    bool dispatchTouchEvent(const PlatformTouchEvent&, const AtomString&, const EventTargetTouchMap&, float, float);
+    void updateTouchLastGlobalPositionAndDelta(unsigned, const IntPoint&, bool, bool);
```
It would be nice to replace the `bool` flags with named enums (even if they're just private to `EventHandler`), for readability.

I think it would also be a bit clearer to use `PointerID` instead of `unsigned` for the first argument.

aprotyas replied:
> It would be nice to replace the bool flags with named enums (even if they're just private to EventHandler), for readability.

Done! https://github.com/WebKit/WebKit/commit/eed593fcf230c5752fc982cef0b7cbdda782897a

> I think it would also be a bit clearer to use PointerID instead of unsigned for the first argument.

Good idea. I'd opted for `unsigned` because that's how `EventHandler` was tracking touch indices in general, but from what I can tell `PointerID == unsigned` so I'll use `PointerID` instead.

whsieh reviewed (approved):
LGTM otherwise!

pxlcoder commented on LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields.html:
```diff
@@ -0,0 +1,61 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <script src="../../../../resources/js-test-pre.js"></script>
```
This can just be `<script src="../../../../resources/js-test.js"></script>`, and you can remove the `js-test-post.js` below.

aprotyas replied:
Thanks, fixed in https://github.com/WebKit/WebKit/commit/eed593fcf230c5752fc982cef0b7cbdda782897a. When do we need to use the `pre`/`post` scripts instead?

pxlcoder replied:
I'm not sure there is ever a need for them over `js-test.js`. My understanding is that they are just the old way of doing things.

aprotyas committed:
movementX and movementY are missing from pointerMove events (255207)
rdar://108112600
https://bugs.webkit.org/show_bug.cgi?id=255207

Reviewed by Wenson Hsieh.

This commit performs a few things:

- Expose `movementX`/`movementY` *generally* in MouseEvent structs
  rather than locking it behind POINTER_LOCK support. This is because
  even without implementing pointer lock on iOS, we can (a) easily
  compute these values and (b) the attributes are generally useful.

- Introduce a bunch of plumbing to pass touch event deltas from our
  EventHandler to the PointerEvent creation. The bookkeeping required to
  compute these touch deltas is placed in WebKitAdditions.

- Tangentially, changes the `PointerID` type to represent uint32_t
  instead of its signed sibling. I've done this because we're passing in
  unsigned identifiers when creating PointerEvents, and I've logged a
  few cases of ID *overflow*.

- Add a layout test to verify correctness of the movementX/Y fields of
  pointer events generated by touch gestures.

- Corrects some test expectations now that we expose
  `movementX`/`movementY` regardless of pointer lock support.

* LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields-expected.txt: Added.
* LayoutTests/fast/events/pointer/ios/drag-populates-pointer-events-with-movementxy-fields.html: Added.
* LayoutTests/platform/gtk/pointer-lock/mouse-event-delivery-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/pointerevents/pointerlock/pointerevent_movementxy_with_pointerlock-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/pointerlock/movementX_Y_basic-expected.txt:
* Source/WebCore/dom/MouseEvent.idl:
* Source/WebCore/dom/MouseEventInit.idl:
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/ios/PointerEventIOS.cpp:
(WebCore::PointerEvent::create):
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::clear):
(WebCore::EventHandler::handleTouchEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::dispatchEventForTouchAtIndex):
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/platform/PointerID.h:
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/263829@main
